### PR TITLE
Fix wx issues from upgrade

### PR DESCRIPTION
**Pull request by Ocraftyone**
Attempting to fix some of the issues that have been introduced by the wxWidgets upgrade

### CODE REVIEW

**Comment by Ocraftyone** [comment]
@Noisyfox I would like to get some help here. I am having some trouble getting SwitchButtons (specifically the objects toggle). I haven't been able to narrow down what is causing the issue. I would assume that it is within the Resize function as that is where it is drawn.

I was able to solve the scaling issues with the topbar, advanced toggle, and about menu. I ended up reversing the change to how the bitmap bundle is generated. It seemed to cause some scaling issues. What I did instead (and what PS did) is allow both the height and width to be specified on both a ScalableBitmap and the get_bmp_bundle function. This seems to have fixed the issue. Not sure why this works better than extrapolating the missing axis dimension like the old function did, but oh well. I may take a look at it later and see if it is worth exploring.

**Comment by Noisyfox** [comment]
Sure. I'll take look soon.

**Comment by Noisyfox** [comment]
There are some other issues such as the highlight color of the window controls are incorrect:
1.8:
image
1.9:
image

**Comment by Ocraftyone** [comment]
@Noisyfox I am not getting the same issue. can you confirm it still exists?

**Comment by Noisyfox** [comment]
> @Noisyfox I am not getting the same issue. can you confirm it still exists?

Windows theme set to light mode and you should get this issue on the Orca's main window.

**Comment by Noisyfox** [comment]
image
Also these icons are misaligned.

**Comment by discip** [comment]
I assume this #2956 is also related.

**Comment by Noisyfox** [comment]
I managed to solve(?) part of the issue scaling https://github.com/Noisyfox/OrcaSlicer/commit/01969879f1c71dc0e3be63634f0beeca4c085ea6  though the size and the position is still incorrect.

**Comment by Noisyfox** [comment]
> image Also these icons are misaligned.

I've fixed this issue as well https://github.com/Noisyfox/OrcaSlicer/commit/d5b9a9be762f7a1d178b3759a8608df0a3af8826

**Comment by Noisyfox** [comment]
image
This part looks odd though.

**Comment by Ocraftyone** [comment]
Awesome! Apologies, I have had to focus on finishing the semester :)

**Comment by tsmith35** [comment]
> There are some other issues such as the highlight color of the window controls are incorrect

The buttons are still light blue with this PR, but not a big deal to me.

I noticed a couple of anomalies with the Device screen Control box:
Device-Control

and Filament options. OrcaSlicer first:
OrcaSlicer Filament Options

Bambu Studio second:
Bambu Studio Filament Options

Overall, looks good. Thanks, @Ocraftyone!

**Comment by tsmith35** [comment]
Okay, last anomaly before bedtime. Device-Control box bottom
AMS display with oversized toolhead

**Comment by Eldenroot** [comment]
Another thing - make "filament color" column for obejct bigger - Bambu studio did some changes to this. It is easier to click on it. Here is the screenshot:

image


Orca slicer does not use the whole cell dimension, also it is a little tricky if you click with mouse not directly on the cell with color selection. I do not if it is was fixed with the latest merge with BL studio or not. I need to download the newest build.

**Comment by Eldenroot** [comment]
Also there is an issue: 
[#](https://github.com/SoftFever/OrcaSlicer/issues/2982)

Add any printer (not Bambulab, better Klipper/Marlin based) and click on the connection settings.

image

**Comment by SoftFever** [comment]
Fixed 
Screenshot 2023-12-21 at 12 37 26 AM

**Comment by SoftFever** [comment]
@Ocraftyone @Noisyfox 
I think we should keep ScalableBitmap's one of the constructor back compatible with previous behave instead of changing codes everywhere. And I'm still not sure how many more to change TBH
e.g. make this following function consistent with old behaviours. I can take a closer look tomorrow.
What's your thought?
ScalableBitmap::ScalableBitmap( wxWindow *parent, 
                                const std::string& icon_name,
                                const int  width/* = 16*/,
                                const int  height/* = -1*/,
                                const bool grayscale/* = false*/):

File: src/slic3r/GUI/wxExtensions.hpp
Comment (SoftFever) [inline src/slic3r/GUI/wxExtensions.hpp:173]: @Ocraftyone @Noisyfox 
FYI. I made this change.
Did a quick scan through the most UI icons, all looks correct on Windows

**Comment by SoftFever** [comment]
> Okay, last anomaly before bedtime. Device-Control box bottom AMS display with oversized toolhead

Fixed in https://github.com/SoftFever/OrcaSlicer/pull/3032/commits/4a01eef6346cecf78dbaaa2fd939a001d9b65193

**Comment by Noisyfox** [comment]
> I think we should keep ScalableBitmap's one of the constructor back compatible with previous behavior instead of changing codes everywhere. And I'm still not sure how many more to change TBH

Agreed

**Comment by Ocraftyone** [comment]
@SoftFever I agree that it seems to be a giant headache to go around an fix all of these issues. Honestly, I don't even exactly see the benefit in the current use case. I don't exactly like leaving it unfinished, but it may just need to be done here.

**Comment by SoftFever** [comment]
> @SoftFever I agree that it seems to be a giant headache to go around an fix all of these issues. Honestly, I don't even exactly see the benefit of using bundles in the current use case. In most cases, only a single bitmap is used. I don't exactly like leaving it unfinished, but it may just need to be done here.

No worries.
You have done a fantastic job already.
With all the changes done in this PR, I see no UI glitches any more.
Well done @Ocraftyone @Noisyfox .

Merged! Cheers

**Comment by SoftFever** [comment]
> Also there is an issue:
> 
> #2982
> 
> Add any printer (not Bambulab, better Klipper/Marlin based) and click on the connection settings.
> 
> image

Thanks.
We will treat it as different issue.

**Review by SoftFever — APPROVED** [review]

**Comment by Ocraftyone** [comment]
Thank you @Noisyfox for finishing this up 😁

**Comment by Noisyfox** [comment]
b3a72efd6e9abfe959c40e546ec25ead
Having this issue at 200% scaling on Windows.

**Comment by tsmith35** [comment]
I'll be honest; I went back to Bambu Studio.

OrcaSlicer devs Ocraftyone and Noisyfox were making great progress, but something changed shortly after, when the changes were supposedly merged with main. The UI is now so badly broken that I only use it for calibrations. And to do so, I use a specific earlier build only, the one with a timestamp of 12/20/23 09:58 (F7ACDDDF) for orca-slicer.exe. No compatibility overrides needed. It's not great, with other graphics left to fix, but it's quite usable.

Anything newer is unusable and just plain ugly without using Windows 10 compatibility mode to try and override graphics anomalies. Even then, only when using Windows 10 scaling provided by Windows "System". Scaling with application or system-enhanced doesn't work. "Application" doesn't fix anything, and "System Enhanced" causes partial (and mostly unusable) windows to appear.

The wxWidgets "upgrade" broke a **_ton_** of stuff. I think the disruption caused by the upgrade was neither expected nor trivial. Quite unfortunate.

**Comment by Eldenroot** [comment]
@tsmith35 - to be honest I do not see any of these "broken" ton of stuff as you reported :)

Orca slicer is superior, print results are better than with Bambu Studio... thats why I use it :)

Keep in mind that this is an open source project, hobby work :) so you can send some fixes for "your" issues and help with development :) maybe better than just crying :)

**Comment by tsmith35** [comment]
> @tsmith35 - to be honest I do not see any of these "broken" ton of stuff as you reported :)
> 
> Orca slicer is superior, print results are better than with Bambu Studio... thats why I use it :)
> 
> Keep in mind that this is an open source project, hobby work :) so you can send some fixes for "your" issues and help with development :) maybe better than just crying :)

What? I run at 4k. I've been a fan of OrcaSlicer since I first saw it. Not crying. I'm an engineer, so I'm just offering my personal assessment of the situation. If it upsets you to hear it, perhaps you could just ignore the comment.

Sounds just like the FB groups. It doesn't affect you, so it must not be an issue. Sure, okay, thanks for sharing that. Merry Christmas if it doesn't offend you and happy holidays otherwise.

**Comment by SoftFever** [comment]
@tsmith35 @Eldenroot  
let's try our best to maintain an open, direct, and yet friendly atmosphere in this community.

I'm taking a look at the UI issues. No worries, we will fix them soon 😉. 
I appreciate all the feedback, especially since both of you are willing to try the dev build and provide feedback. 
As a maintainer, I can't express enough how helpful that is!

@tsmith35, your feedback is valid and constructive. I'm the one to blame here. Nevertheless, we are learning.

Lastly, and most importantly, let's enjoy our hobby and Merry Christmas!

**Comment by tsmith35** [comment]
> Lastly, and most importantly, let's enjoy our hobby and Merry Christmas!

Thanks for that, and thanks for the many hours you guys have spent on this project. I enjoy everything you've done to advance OrcaSlicer (and, indirectly, Bambu Studio), something which has improved everybody's user experience. If not for this fork, we would be deprived of many features that we've grown to depend on.

I make my living with industrial control systems and mechanical engineering. It's fun and rewarding, but I depend on blunt, accurate feedback to get my job done. I save my polite, somewhat ambiguous speaking for dealing with management. lol

I wish you all a very merry Christmas, and I look forward to trying every new release. Thanks, guys.

**Pull request merged**
